### PR TITLE
tests: Introduce testing dialect

### PR DESCRIPTION
**Pull request by webmiche**
This PR introduces the testing dialect. This is a dialect that allows use to write filecheck tests without introducing unnecessary dependencies to other dialects. In particular, a lot of our tests depend on the func dialect, which we want to keep in sync with MLIR's version. If they ever change it, we need to refactor a lot of tests just because of this non-fundamental dependency. For exmaple, they moved the func operation from builtin to the (at that point new) func dialect, which mmeant we had to rename all `builtin.func` to `func.func`. This is the kind of refactoring I want to prevent.

 At the moment, this dialect contains an operation that produces (an arbitrary amount of) arbitrarily typed results and a testing type. This type is parametric and is supposed to be used in cases where the constraint is `AnyAttr`. In cases that have actual constraints (like for example `SignLessIntegerLike` in arith) we will need to keep using that. I argue that this kind of dependency is really a fundamental one, so I don't think we need to circumvent this. The one operation that I don't think we can circumvent atm is the builtin.module, but I don't expect that to change too much so I think we are fine.

For the time being, this only shows two example rewritten tests. Let me know if you like the design. I will in the meantime rewrite the other tests.

### CODE REVIEW

**Review by georgebisbas — COMMENTED** [review]
To be honest, I am not sure something like this is useful, but rather adds complexity instead of removing it. But maybe it is only my personal opinion. A dialect for testing is not something I see as useful. But I am curious to see what others think.

File: tests/filecheck/func_ops.xdsl [hunk 1]
```diff
@@ -15,21 +15,21 @@ builtin.module() {
...
+  // CHECK: func.func() ["sym_name" = "call_void",
   // CHECK-NEXT:  func.call() ["callee" = @call_void]
   // CHECK-NEXT:  func.return()
   // CHECK-NEXT:}
 
-  func.func() ["sym_name" = "arg_rec", "function_type" = !fun<[!i32], [!i32]>, "sym_visibility" = "private"] {
-  ^0(%0 : !i32):
-    %1 : !i32 = func.call(%0 : !i32) ["callee" = @arg_rec]
-    func.return(%1 : !i32)
+  func.func() ["sym_name" = "arg_rec", "function_type" = !fun<[!test.type<"int">], [!test.type<"int">]>, "sym_visibility" = "private"] {
```
Comment (georgebisbas) [inline tests/filecheck/func_ops.xdsl:23]: Assuming I am reading this filecheck test, why should I need to go to test dialect to understand what is going on?

Comment (webmiche) [reply tests/filecheck/func_ops.xdsl:23]: You mean when a new user wants to understand how IRs look like or how the func dialect looks like?

File: xdsl/dialects/test.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from __future__ import annotations
+from xdsl.ir import Attribute, Data, Dialect, MLIRType, OpResult, Operation
+
+from xdsl.irdl import VarOpResult, irdl_attr_definition, irdl_op_definition
+from xdsl.parser import BaseParser
+from xdsl.printer import Printer
+
+
+@irdl_op_definition
+class ProduceValuesOp(Operation):
```
Comment (math-fehr) [inline xdsl/dialects/test.py:10]: Can we add a comment here to explain how this is meant to be used?
That should make it clear for new users.

File: xdsl/dialects/test.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    @staticmethod
+    def get_values(
+            *res: Attribute) -> tuple[ProduceValuesOp, tuple[OpResult, ...]]:
+        op = ProduceValuesOp.from_result_types(*res)
+        return op, tuple(op.results)
+
+
+@irdl_attr_definition
+class TestType(Data[str], MLIRType):
```
Comment (math-fehr) [inline xdsl/dialects/test.py:27]: Same here.

**Review by math-fehr — APPROVED** [review]
Nice!
For answering @georgebisbas, I feel that removing most dependencies in favor to the test dialect should not confuse too much newcomers, compared to what we currently have.

I expect the test dialect to have only a few operations (maybe we should add test.func?), and reducing the dependencies on other MLIR dialects should avoid us some really annoying refactors. Also, if users want to use different MLIR versions, but still use some of our dialects, then this should allow them to reuse our tests, since I expect the test dialect to be more stable than the MLIR dialects.

File: tests/filecheck/func_ops.xdsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (webmiche) [reply tests/filecheck/func_ops.xdsl:23]: I am not sure I understand your question. Notice that I could pass in test.type<"any string">, but that the type need to stay the same as I am recursively calling the function itself.

**Comment by AntonLydike** [comment]
I am unsure about the utility that `test.type` provides. My fear is that it makes the code pretty unreadable. I'd rather have a sensible example than one where `i32` is replaced by `test.type<"int">`, because I feel that the former is much more readable and clear.

**Comment by webmiche** [comment]
The issue with `i32` is that it is a builtin type and needs to be held uptodate with MLIR to allow interoperation. Therefore, the whole argument for the test dialect extends to that attribute, IMO.

Do you think there could be a better name? I can rename to just `type<"a">`, if you feel that's better. Also notice that I am not trying to write nice tests here, I am trying to write sensible tests that do not have the MLIR problem.

**Comment by tobiasgrosser** [comment]
Just my two cents. My personal feeling is that using the most foundational dialects from MLIR as test dialect is likely the most straightforward way. I doubt that i32 changes in the future. Similar to how func likely remains constant. In case these dialects change, we anyhow need to update our test cases. I feel we might be over-engineering this aspect a little. However, you guys have probably more direct experience with this issue.

**Comment by AntonLydike** [comment]
I agree with tobias here, I think there are a couple of things in builtin that are pretty safe to stay the way they are:

 - most integer and float types, especially `i*` and `f32`/`f64` should be safe 
 - `builtin.module`
 - `func.func` in its basic form (without all the extra attributes)
 - SymbolRefAttr (the `@ref` thingies)
 - ArrayAttr
 - The FunctionType e.g. `(arg types) -> (out types)`
 - Basic ranked memref (and probably unranked memref)
 - Vector

Probably some more stuff that I'm forgetting

I think the priority should be in reducing dependencies on operations and types from other dialects. The `builtin` stuff should be relatively safe as long as you stay in the basics and aren't doing anything super special, or am I too naive?

**Comment by webmiche** [comment]
While I don't think I have the necessary insight and experience to do strong predictions, I do know that I have lived through one major refactor (splitting std) which was quite painful ([here](https://github.com/xdslproject/xdsl/pull/32/files#diff-d386c4a1ac37bbdeaf300353e3c78b90a57a2517a2ba6bf07446875d2f8c79a7) the PR).

This was at a time where we supported just a very small set of MLIR operations (only things used for integer arithmetic in std plus a bit more) and this was already a 500 line PR. Furthermore, we did not have a MLIR-based CI, there were three users of the repo, which were also the main engineers working on the PR, so noone was blocked, and we only had very minor testing. Now that the repo has evolved, this will look very different. The most recent [update](https://github.com/xdslproject/xdsl/pull/327) was another 600 line PR, that fortunately did not touch the core dialects a lot.

I also do know that there is fundamental work on the core structures going on in MLIR (like Properties), but I don't know how much that will touch us in the future (that particular one probably won't at all). 

> I think the priority should be in reducing dependencies on operations and types from other dialects. 

I agree with you and the list, but I think what we need is a uniform but strong testing framework (for both filecheck and pytest). And that is something that I think we should make as little dependent on MLIR-based dialects as possible. This is only the testing framework for filecheck, I am thinking about how to do something similar in pytest as well.

Let me ask you then: What do we lose with the testing dialect? Is it just he added complexity of one more dialect?

From my point of view, we are losing very little with the testing dialect and I am personal ready to pay this to get testing that is less dependent on outside dependencies like MLIR.

**Comment by math-fehr** [comment]
I'm more in the side with @webmiche on that.

Updating MLIR version is not really fun, and things we think are safe (builtin types, and operations), are often not. That's why I would like to rely as less as possible on MLIR dialects. There is always a refactor in MLIR that ends up being a problem in our code. The first time was the split of the `std` dialect, last time was the change of attribute for the `operand_segment_sizes` attributes, and who knows what's next, especially since one of the big change that I'm planning on MLIR is adding names for types and attributes.

Even in xDSL, once we are going to start looking at custom parsing and printing, this is going to require a large refactor as well. If we can concentrate these refactors as much as we can, that would be good, because otherwise we would just be losing a lot of time on these.

Besides all of this, I doubt that we are losing that much by going with a test dialect. Yes, we have to write `!test.type<"a">` instead of `i32`, but this is quite minimal. Also, note that this would only be for unit tests, not for integration tests, which would present examples, and lowering examples.

**Review by superlopuh — APPROVED** [review]

File: xdsl/xdsl_opt_main.py [hunk 4]
```diff
@@ -188,6 +189,7 @@ def register_all_dialects(self):
         self.ctx.register_dialect(LLVM)
         self.ctx.register_dialect(Vector)
         self.ctx.register_dialect(GPU)
+        self.ctx.register_dialect(Test)
```
Comment (AntonLydike) [inline xdsl/xdsl_opt_main.py:192]: We might want to gate the testing dialect behind a special flag? I am unsure if we want it in the general `xdsl-opt` if that makes sense?

**Review by AntonLydike — APPROVED** [review]
I'll side with @math-fehr and @webmiche on this, as you have much more experience and your comments do make a lot of sense to me.

I just have a minor nitpick that we might not want to have the testing dialect enabled by default, but that's just a random thought, nothing concrete.

**Review by PapyChacal — APPROVED** [review]
I'm in doubt about the actual benefits too. While I can't argue about the stability of MLIR to come, I agree with Tobias on one point:

> In case these dialects change, we anyhow need to update our test cases.

But then again, you are the one who went through the motivating experience and did the effort, so I say let's just try. Would be nice to have your feedback on it making a difference or not if we face one of those heavy refactors!

File: xdsl/xdsl_opt_main.py (hunk 4 above)
Comment (webmiche) [reply xdsl/xdsl_opt_main.py:192]: I must say I am not sure at all and I had to think about this quite a lot to actually figure out an opinion. This is a very very weak opinion, but I don't feel like this is necessary. I don't think we lose anything from making it available per default and we add more complexity to the lit tests if we add a flag.

I also think there is value in tests being able to run from the command line without needing more than a few flags. Definitely a good and justified suggestions though! :)

File: tests/filecheck/func_ops.xdsl (hunk 1 above)
Comment (webmiche) [reply tests/filecheck/func_ops.xdsl:23]: @georgebisbas gentle ping

[the remark this answers is not in the document]
Comment (webmiche) [reply tests/filecheck/func_ops.xdsl:23]: @superlopuh gentle ping :)

Comment (georgebisbas) [reply tests/filecheck/func_ops.xdsl:23]: I was thinking of how understandable is a test from someone that has not seen the test dialect

**Comment by georgebisbas** [comment]
I am happy to approve after reading all your comments. I see consensus and I trust you that this towards the right direction.

**Review by georgebisbas — APPROVED** [review]

File: tests/filecheck/func_ops.xdsl (hunk 1 above)
Comment (webmiche) [reply tests/filecheck/func_ops.xdsl:23]: It isn't and I don't think it's supposed to be. IMO, people should start with the documentation. If there is no docu, they will go look at test cases (at least that's how I am doing it). So, I think, we should not design our testing with new user readability in mind.

File: xdsl/xdsl_opt_main.py (hunk 4 above)
Comment (AntonLydike) [reply xdsl/xdsl_opt_main.py:192]: Yeah, I think you are right. This point will become moot anyway once our tests aren't the best source of documentation I guess :D

**Comment by webmiche** [comment]
closes #203

**Comment by webmiche** [comment]
closing this until we find stronger consensus (or the next refactor comes up)

File: tests/filecheck/func_ops.xdsl (hunk 1 above)
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/filecheck/func_ops.xdsl:23]: ah ok, I see, all good

**Pull request closed without merging**